clients/src/main/java/org/apache/kafka/clients/producer/internals/BufferPool.java

```diff
@@ -128,11 +128,12 @@ public ByteBuffer allocate(int size, long maxTimeToBlockMs) throws InterruptedEx
                     long timeNs = Math.max(0L, endWaitNs - startWaitNs);
                     this.waitTime.record(timeNs, time.milliseconds());
 
-                    if (waitingTimeElapsed)
+                    if (waitingTimeElapsed) {
+                        this.waiters.removeLast();
```

Comment:
No, it's fine to update this PR to use `waiters.remove(moreMemory)`. What I meant was that the existing unit test passes with either implementation (as I understand) so it's perhaps too simple. It would be good if it actually failed if `waiters.removeLast` was used. Does that make sense?